### DISCUSSION TITLE
How to make an Ajax call to a php function in my plugin

### DISCUSSION THREAD

**Discussion by ferdware** [givanz/Vvveb#208] [General]
I created and added a Plugin. Then I added a test button on the Product.html page with an Onclick event to a function in app.js.  What I'd like to do is to make an Ajax call to a php function that I have in my plugin?

**Comment by givanz** [comment]
The easiest way to call php code is to create a new controller in your plugin.

`plugins/test-plugin/app/controller/index.php`

```php
namespace Vvveb\Plugins\TestPlugin\Controller;

use Vvveb\Controller\Base;

class Index extends Base {
	function index() {
		// this will try to render /plugins/test-plugin/public/index.html 
		// using the /plugins/test-plugin/app/template/index.tpl template
	}

	function json() {
		$this->response->setType('json');
		
		return ['key' => 'value'];
		// or
		//$this->response->output(['key' => 'value']);
	}
}
```

You can then call the method with [localhost/?module=plugins/test-plugin/index&action=json](localhost/?module=plugins/test-plugin/index&action=json)

Or you can register a new route with `addRoute`
```php
use Vvveb\System\Routes;
class TestPlugin {

	function app() {
		//add new route for plugin page
		Routes::addRoute('/test-plugin',  ['module' => 'plugins/test-plugin/index/json']);
	}
}
```
and then you can call [localhost/test-plugin](localhost/test-plugin)

You can quickly check the example from the included test plugin `/plugins/test-plugin`

**Comment by ferdware** [comment]
Thank You so much Givanz!

 

I have 15 websites currently on Shopify that I eventually want to transition over to your solution.  I also want to become a big promoter of your system too. I think what you’re doing is AMAZING!  I just need to understand the workings so I know how to customize for the needs of my clients. 

 

Thank You for your work! I really appreciate what you’re doing!

 

Ferdie Eugenio

 

From: Givan ***@***.***>
Reply-To: givanz/Vvveb ***@***.***>
Date: Monday, October 28, 2024 at 1:41 PM
To: givanz/Vvveb ***@***.***>
Cc: ferdware ***@***.***>, Author ***@***.***>
Subject: Re: [givanz/Vvveb] How to make an Ajax call to a php function in my plugin (Discussion #208)

 

The easiest way to call php code is to create a new controller in your plugin.

plugins/test-plugin/app/controller/index.php

namespace Vvveb\Plugins\TestPlugin\Controller;
 
use Vvveb\Controller\Base;
 
class Index extends Base {
        function index() {
               // this will try to render /plugins/test-plugin/public/index.html 
               // using the /plugins/test-plugin/app/template/index.tpl template
        }
 
        function json() {
               $this->response->setType('json');
               
               return ['key' => 'value'];
               // or
               //$this->response->output(['key' => 'value']);
        }
}
You can then call the method with localhost/?module=plugins/test-plugin/index&action=json

Or you can register a new route with addRoute

use Vvveb\System\Routes;
class TestPlugin {
 
        function app() {
               //add new route for plugin page
               Routes::addRoute('/test-plugin',  ['module' => 'plugins/test-plugin/index/json']);
        }
}
and then you can call localhost/test-plugin

You can quickly check the example from the included test plugin /plugins/test-plugin

**Comment by ferdware** [comment]
I noticed there is two folders for “Plugins”

 

/plugins/

 

/public/plugins

 

Which one do I edit to make changes?

 

Ferdie Eugenio 

 619-857-8098

***@***.***

https://kreativeportals.com

**Comment by givanz** [comment]
Thank you for using and promoting Vvveb.

In Vvveb if securely installed only `/vvveb/public/` folder is available for external access, no php file except `vvveb/public/index.php` can be accessed. 

Plugins are stored in `/vvveb/plugins/` folder and no plugin file can be accessed directly.

To overcome this issue for plugin files that need to be publicly accessible, when activating a plugin for the first time all files inside `/vvveb/plugins/my-plugin/public` are copied to `/vvveb/public/my-plugin` 

Plugin files inside `/vvveb/plugins/my-plugin/public` folder should only include public files such as html templates, plugin images/icon, css, js files or other static files or assets.

For an active plugin static files are loaded from `/vvveb/public/my-plugin` files are only copied once when the plugin is activated first time and the files are not synced afterwards.

After finishing development you should copy the changes from `/public/my-plugin/` back to `/plugins/my-plugin/public` and zip the folder for plugin distribution.
